app/models/grda_warehouse/auth_policies/user_acl_context.rb

```diff
@@ -32,18 +35,53 @@ def initialize(user)
     permissions_for_collection_ids(collection_ids)
   end
 
+  def preload_project_dependencies(project_ids)
+    preload_coc_codes_by_project(project_ids)
+    preload_collection_ids_by_project(project_ids)
+  end
+
   # Duck-typed for legacy role-based permissions
   def legacy_permissions
     EMPTY_SET
   end
 
   protected
 
+  def preload_coc_codes_by_project(project_ids)
+    p_t = GrdaWarehouse::Hud::Project.arel_table
+    results = GrdaWarehouse::Hud::ProjectCoc.
+      joins(:project).
+      where(p_t[:id].in(project_ids)).
+      pluck(p_t[:id], :coc_code).
+      group_by(&:shift).
+      transform_values(&:flatten)
+    @coc_codes_by_project.merge!(results)
+  end
+
+  def coc_codes_for_project(project_id)
+    preload_coc_codes_by_project([project_id]) unless @coc_codes_by_project.key?(project_id)
+    @coc_codes_by_project[project_id] || []
+  end
+
+  def preload_collection_ids_by_project(project_ids)
+    results = GrdaWarehouse::ProjectCollectionMember.
+      where(project_id: project_ids).
+      pluck(:project_id, :collection_id).
+      group_by(&:shift).
+      transform_values(&:flatten)
+    @collection_ids_by_project.merge!(results)
+  end
+
+  def collection_ids_for_project(project_id)
+    preload_collection_ids_by_project([project_id]) unless @collection_ids_by_project.key?(project_id)
+    @collection_ids_by_project[project_id] || []
```

Comment:
Looking at this more closely, I think the null result isn't cached. (preload_collection_ids_by_project wouldn't have ids for projects without collections). So you could cache the null result here (same for coc_codes_for_project and equiv methods in legacy context) to avoid n+1s in that case
```suggestion
    @collection_ids_by_project[project_id] ||= []
```